### PR TITLE
Fix case-sensitive semantics of some catalog columns

### PR DESCRIPTION
**Pull request by skumawat2025**
### Description

Fix case-sensitive semantics of some catalog columns - 
some sys catalogs listed below have default collation of postgres because of wrong data-type (pg_catalog) hence the default collation of these catalog columns shows case-sensitive behaviour which shouldn't be the case in Babelfish. 

1. sys.stats -  changed datatype of stats_generation_method_desc to sys.varchar(255) (previously it was varchar(255) )
2. sys.dm_exec_connections- 
    1. column local_net_address is of type varchar(48), changing it to sys.varchar(48) (Testing is tricky here. As this stores string of address like 172.31.68.239 or NULL). 
    2. column client_net_address is of type varchar(48), changing it to sys.varchar(48)
    3. column endpoint_id type changed to int. 
3. sys.syscolumns -  column printfmt datatype is varchar(255) changing it to sys.varchar(255)
    1. column printfmt datatype is varchar(255) changing it to sys.varchar(255)
    2. column type and xtype should be tinyint according to tsql but doing that in bbf results in overflow. So we are not changing it. 
4. sys.system_objects - In this catalog, in babelfish we are returning some extra number of columns (oid, nspname, nspowner, nspacl). The column nspname shows unexpected collation. Fixed these. 
5. sys.xml_indexes - Changing datatype of column secondary_type to sys.bpchar(1) from char(1).
6. sys.sysforeignkeys - In this catalog we have extra column name in babelfish that is not there in tsql. So we need to remove this apart from that object_id name should be changed to constid.  (removed name column and constid, fkeyid, rkeyid changed to int from oid)
7. sys.data_spaces - Fixed this catalog by changing datatype of column type to sys.bpchar
8. sys.sysprocesses-
     1. spid, and blocked columns should be smallint, but changing in babelfish will result in overflow. Not changing these.
     2. kpid (Changing to smallint) 
     3. waittype (changing to sys.binary(2))
     4. waittime (changing to bigint)
     5. lastwaittype  (Changing it to sys.nchar(32)) (Can not test as it is always NULL)
     6. waitresource  (Changing it to sys.nchar(256))(Can not test as it is always NULL)
     7. dbid changed to int, ecid changed to smallint
     8. cpu (changing to int)
     9. physical_io (changing to bigint)
     10. memusage (changing to int)
     11. login_time (changing to sys.datetime)
     12. last_batch (changing to sys.datetime)
     13. open_tran (changing to smallint)
     14. status (Changing it to sys.nchar(30))
     15. sid (changing to sys.binary(86))
     16. program_name  (Changing it to sys.nchar(128))
     17. hostprocess  (Changing it to sys.nchar(10))
     18. cmd (Changing it to sys.nchar(52));
     19. nt_domain (Changing it to sys.nchar(128))
     20. nt_username (Changing it to sys.nchar(128))
     21. net_address  (Changing it to sys.nchar(12))
     22. net_library (Changing it to sys.nchar(12))
     23. loginname (Changing it to sys.nchar(128))
     24. context_info column should be binary() type currently it's bytea type because of upgrading issue.
     25. sql_handle (changing to sys.binary(20))
     26. stmt_start, stmt_end, request_id (changing to int)

10. sys.key_constraints - data-type of column type changed it to sys.bpchar(2) from char(2)
11.  sys.foreign_keys - 
      1. Changed datatype of column type to sys.bpchar(2) from char(2). 
12. sys.views - Fixed the following columns data-type 
     1. name column data-type changed to sysname
     2. type column data-type changed to sys.bpchar(2)
     3. type_desc column data-type changed to sys.nvarchar(2)
     4. create_date column data-type changed to sys.datatime
     5. modify_date column data-type changed to sys.datetime
     6. is_ms_shipped, is_published, is_schema_published, with_check_option, is_date_correlation_view, is_tracked_by_cdc columns data-type changed to sys.BIT.
     7. object_id, schema_id changed to int from oid
13. sys.tables - type column data-type changed to sys.bpchar(2)
14. sys.default_constraints - Changed datatype of column type from char(2) to sys.bpchar(2)
15. sys.check_constraints - Changed datatype of column type from char(2) to sys.bpchar(2). 
16.  sys.types - 
      1. name column was of type pg_catalog.text changed it to sys.sysname
      2. collation_name column was of type pg_catalog.text changed it to sysname
      4. precision, scale column was of type int changed it to tinyint
      5. is_user_defined, is_assembly_type, is_table_type Changed it to sys.BIT
      6. default_object_id, rule_object_id changed it to int. 
      7. system_type_id, user_type_id, schema_id  changed to int. 
17. sys.sp_sproc_columns_view - This view is dependent on sys.types hence below columns data-type should be changed.
      1. PRECISION, SCALE column data-type changed to tinyint -> these changed reverted back to int for precision and smallint for scale to match with tsql. 
18. sys.systypes - This view is dependent on sys.types hence below columns data-type should be changed.
      1. In prec column data-type precision should cast to smallint as it got changed to tinyint in sys.types.
   
sys.types is dependency of sys.table_types and sys.all_parameters so some respective changes also done in these catalogs. 

Catalog Columns list with wrong datatype and unexpected collation - 
```
select count(attname) attname_count, relname from ((pg_attribute attr join pg_class rel on attr.attrelid = rel.oid) join pg_type typ on attr.atttypid = typ.oid) join pg_namespace nsp on nsp.oid = typ.typnamespace  where rel.relnamespace = (select oid from pg_namespace where nspname = 'sys') and (attcollation = 100 or attcollation = 950) and atttypid > 0 GROUP BY relname;
```

 ```
  attname_count |                 relname                 
---------------+-----------------------------------------
             1 | database_principals
             2 | types
             6 | babelfish_syslanguages
            15 | babelfish_configurations_view
             1 | tables
             1 | server_principals
             8 | sp_columns_100_view
             2 | shipped_objects_not_in_sys
             2 | babelfish_namespace_ext
             3 | babelfish_extended_properties_pkey
             6 | spt_datatype_info_table
             1 | stats
             1 | babelfish_server_options
            11 | sysprocesses
             1 | system_objects
             3 | views
             2 | babelfish_helpcollation
             1 | data_spaces
             3 | service_settings
             2 | versions
             2 | pg_namespace_ext
            13 | spt_columns_view_managed
             1 | babelfish_namespace_ext_pkey
             1 | babelfish_view_def
             2 | babelfish_authid_login_ext
             1 | babelfish_authid_login_ext_pkey
             1 | babelfish_server_options_pkey
             1 | syscolumns
             1 | babelfish_authid_user_ext_login_db_idx
             4 | babelfish_sysdatabases
             5 | babelfish_function_ext
             1 | sysdatabases
             1 | xml_indexes
             2 | babelfish_function_ext_pkey
             1 | babelfish_authid_user_ext_pkey
             1 | check_constraints
             1 | key_constraints
             4 | babelfish_has_perms_by_name_permissions
             3 | babelfish_authid_user_ext
             2 | dm_exec_connections
             1 | babelfish_sysdatabases_pkey
             4 | sp_tables_view
             1 | sysforeignkeys
             3 | babelfish_extended_properties
             2 | table_types
             1 | foreign_keys
             1 | default_constraints
(47 rows)
```

After the fixes done in this PR - 
```
attname_count |                 relname                 
---------------+-----------------------------------------
             2 | babelfish_authid_login_ext
             1 | babelfish_authid_login_ext_pkey
             3 | babelfish_authid_user_ext
             1 | babelfish_authid_user_ext_login_db_idx
             1 | babelfish_authid_user_ext_pkey
            15 | babelfish_configurations_view
             3 | babelfish_extended_properties
             3 | babelfish_extended_properties_pkey
             5 | babelfish_function_ext
             2 | babelfish_function_ext_pkey
             4 | babelfish_has_perms_by_name_permissions
             2 | babelfish_helpcollation
             2 | babelfish_namespace_ext
             1 | babelfish_namespace_ext_pkey
             1 | babelfish_server_options
             1 | babelfish_server_options_pkey
             4 | babelfish_sysdatabases
             1 | babelfish_sysdatabases_pkey
             6 | babelfish_syslanguages
             1 | babelfish_view_def
             1 | database_principals
             2 | pg_namespace_ext
             1 | server_principals
             3 | service_settings
             2 | shipped_objects_not_in_sys
             8 | sp_columns_100_view
             4 | sp_tables_view
            13 | spt_columns_view_managed
             6 | spt_datatype_info_table
             1 | sysdatabases
             2 | versions
(31 rows)

```
Signed-off-by:  Sandeep Kumawat <skumwt@amazon.com>

### Issues Resolved
BABEL-4529

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 1]
```diff
@@ -987,8 +985,7 @@ GRANT SELECT ON sys.procedures TO PUBLIC;
 
 create or replace view sys.sysforeignkeys as
 select
-  c.conname as name
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:990]: what would happen if some cx dependent objects are there for this column?

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 2]
```diff
@@ -1043,8 +1040,8 @@ select
   , coalesce(blocking_activity.pid, 0) as blocked
   , null::bytea as waittype
   , 0 as waittime
-  , a.wait_event_type as lastwaittype
-  , null::text as waitresource
+  , CAST(a.wait_event_type as sys.nchar(32)) as lastwaittype
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:1045]: since we are fixing this then lets make sure that data type of all columns matches with TSQL

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 3]
```diff
@@ -1674,7 +1671,11 @@ left join sys.shipped_objects_not_in_sys nis on nis.name = ('TT_' || tt.name ||
 GRANT SELECT ON sys.all_objects TO PUBLIC;
 
 create or replace view sys.system_objects as
-select * from sys.all_objects o
+select
+  name, object_id, principal_id, schema_id, 
+  parent_object_id, type, type_desc, create_date, 
+  modify_date, is_ms_shipped, is_published, is_schema_published
+ from sys.all_objects o
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:1680]: why this change?

**Review by Deepesh125 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 1 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:990]: While upgrading we do rename this object. Hence the already dependent  cx object will continue working fine.

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 2 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:1045]: Noted.

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 3 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:1680]: in t-sql sys.system_objects only these columns are supported. In BBF there are some extra columns. So did this change for sys.system_objects.

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql [hunk 4]
```diff
@@ -2669,7 +2669,7 @@ CASE
   WHEN st.is_table_type = 1 THEN 0
   ELSE st.precision 
 END 
-AS sys.int) AS PRECISION
+AS sys.tinyint) AS PRECISION
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: Why do we need this change?

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql [hunk 5]
```diff
@@ -2693,7 +2693,7 @@ CASE
   WHEN st.is_table_type = 1 THEN NULL
   ELSE st.scale
 END
-AS smallint) AS SCALE
+AS sys.tinyint) AS SCALE
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: Why do we need this change?

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 6]
```diff
@@ -18,7 +18,7 @@ select
   , CAST(NULL as int) as principal_id
   , CAST(t.relnamespace  as int) as schema_id
   , 0 as parent_object_id
-  , CAST('U' as CHAR(2)) as type
+  , CAST('U' as sys.bpchar(2)) as type
```
Comment (ahmed-shameem) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:21]: Why can't we use sys.char? I believe char(n) and bpchar(n) are same.

File: test/JDBC/expected/BABEL-2829.out [hunk 7]
```diff
@@ -5,32 +5,32 @@
 SELECT DISTINCT db_name(dbid), loginname FROM sys.sysprocesses WHERE spid = @@SPID
 GO
 ~~START~~
-nvarchar#!#varchar
-master#!#jdbc_user
+nvarchar#!#nchar
```
Comment (ahmed-shameem) [inline test/JDBC/expected/BABEL-2829.out:8]: What about upgrade schedule file?

**Review by ahmed-shameem — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql (hunk 4 above)
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: +1, Please check with T-SQL.

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql (hunk 5 above)
Comment (sumitj824) [reply contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: +1, Please check with T-SQL.

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 8]
```diff
@@ -987,8 +985,7 @@ GRANT SELECT ON sys.procedures TO PUBLIC;
 
 create or replace view sys.sysforeignkeys as
 select
-  c.conname as name
-  , c.oid as object_id
+  c.oid as constid
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_views.sql]: All the datatypes in this view are inconsistent with T-SQL. I will suggest you to check all the system views for datatype inconsistency and prepare notes of that, it will also help us reviewing the PR.

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql (hunk 5 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: So we fixed the data-type of scale and precision of sys.type view to match it with t-sql. And we use sys.types to create sp_sproc_columns_view views. Hence this change .

File: contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql (hunk 4 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/babelfishpg_tsql.sql]: So we fixed the data-type of scale and precision of sys.type view to match it with t-sql. And we use sys.types to create sp_sproc_columns_view views. Hence this change .

**Review by sumitj824 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 6 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:21]: sys.char(n) data-type is not directly accessible via pg-end point. sys.bpchar(n) is same as sys.char(n) in t-sql endpoint.

**Review by ahmed-shameem — APPROVED** [review]
LGTM.

File: contrib/babelfishpg_tsql/sql/sys_views.sql [hunk 9]
```diff
@@ -1182,7 +1181,7 @@ END;
 $$ LANGUAGE plpgsql IMMUTABLE STRICT;
 
 CREATE OR REPLACE VIEW sys.systypes AS
```
Comment (sumitj824) [inline contrib/babelfishpg_tsql/sql/sys_views.sql:1183]: Definition in upgrade is different from this one. Please make sure if we are making some changes in later commit, it should reflect in upgrade as well.

**Comment by sumitj824** [comment]
apart from that, LGTM.

**Review by sumitj824 — APPROVED** [review]

File: contrib/babelfishpg_tsql/sql/sys_views.sql (hunk 9 above)
Comment (skumawat2025) [reply contrib/babelfishpg_tsql/sql/sys_views.sql:1183]: done

**Review by Deepesh125 — APPROVED** [review]
